packages/botonic-react/src/util/webchat.js

```diff
@@ -0,0 +1,151 @@
+
+import { Button } from '../components/button'
+import { WEBCHAT } from '../constants'
+import { getProperty, strToBool } from './objects'
+import { deepMapWithIndex } from './react'
+
+/**
+ * Returns the value of a property defined in bot's theme based on WEBCHAT.CUSTOM_PROPERTIES dictionary.
+ * It gives preference to nested defined properties (e.g.: header.style) over plain properties (e.g.: headerStyle).
+ * If property doesn't exist, returns the defaultValue.
+ */
+export const _getThemeProperty = theme => (
+  property,
+  defaultValue = undefined
+) => {
+  for (const [k, v] of Object.entries(WEBCHAT.CUSTOM_PROPERTIES)) {
+    if (v == property) {
+      const nestedProperty = getProperty(theme, v)
+      if (nestedProperty !== undefined) return nestedProperty
+      const plainProperty = getProperty(theme, k)
+      if (plainProperty !== undefined) return plainProperty
+      return defaultValue
+    }
+  }
+  return undefined
+}
+
+export class ButtonsDisabler {
+  static constructBrowserProps(props) {
+    const disabledProps = { disabled: props.disabled }
+    if (props.autodisable !== undefined)
+      disabledProps.autodisable = strToBool(props.autodisable)
+    if (props.disabledstyle !== undefined)
+      disabledProps.disabledstyle = props.disabledstyle
+    return disabledProps
+  }
+  static constructNodeProps(props) {
+    const disabledProps = {}
+    if (props.autodisable !== undefined)
+      disabledProps.autodisable = String(props.autodisable)
+    if (props.disabledstyle !== undefined)
+      disabledProps.disabledstyle = JSON.stringify(props.disabledstyle)
+    return disabledProps
+  }
+
+  static withDisabledProps(props) {
+    return {
+      ...{ disabled: props.disabled },
+      ...{ autodisable: props.autodisable },
+      ...{ disabledstyle: props.disabledstyle },
+    }
+  }
+
+  static resolveDisabling(theme, props) {
+    const getThemeProperty = _getThemeProperty(theme)
+    const autoDisable =
+      props.autodisable !== undefined
+        ? props.autodisable
+        : getThemeProperty(
+            WEBCHAT.CUSTOM_PROPERTIES.buttonAutoDisable,
+            WEBCHAT.DEFAULTS.BUTTON_AUTO_DISABLE
+          )
+    const disabledStyle =
+      props.disabledstyle !== undefined
+        ? props.disabledstyle
+        : getThemeProperty(
+            WEBCHAT.CUSTOM_PROPERTIES.buttonDisabledStyle,
+            WEBCHAT.DEFAULTS.BUTTON_DISABLED_STYLE
+          )
+    return { autoDisable, disabledStyle }
+  }
+
+  static updateChildrenButtons(children, additionalProps = undefined) {
+    return deepMapWithIndex(children, n => {
+      if (n.type === Button) return this.updateButtons(n, additionalProps)
+      if (n.props && n.props.children) {
+        return {
+          ...n,
+          ...{
+            props: {
+              ...n.props,
+              ...{ children: deepMapWithIndex(n.props.children, n => n) },
+            },
+          },
+        }
+      }
+      return n
+    })
+  }
```

Comment:
I think this implementation is way more complex than it should be. You should never mess directly with React internals, unless there's a very good reason, which I fail to see here.
Why can't we just have these props in our Webchat reducer `messagesJSON` which is just a plain and simple JSON representation of our components? Then have the Button components react to changes on that object?

Reply:
As discussed we agreed to move forward, but it's true that Webchat code must be rethinked in order to decouple all the functionalities in a good way.